cpp/arcticdb/storage/storage.hpp

```diff
@@ -27,6 +27,8 @@ namespace arcticdb::storage {
 
 struct StorageBase {}; // marker class for type checking
 
+using ReadVisitor = std::function<void(const VariantKey&, Segment)>;
```

Comment:
I agree that we must avoid copy, for now this allows compiling.

Specifying a lvalue reference or a rvalue reference for `Segment` does not allow compiling since in some context `Segment` are passed by value or by reference.

<details>
<summary>Details</summary>

For instance, when we use r-value reference:

```diff
diff --git i/cpp/arcticdb/storage/storage.hpp w/cpp/arcticdb/storage/storage.hpp
index 48c71e7..4505672 100644
--- i/cpp/arcticdb/storage/storage.hpp
+++ w/cpp/arcticdb/storage/storage.hpp
@@ -27,7 +27,7 @@ namespace arcticdb::storage {
 
 struct StorageBase {}; // marker class for type checking
 
-using ReadVisitor = std::function<void(const VariantKey&, Segment)>;
+using ReadVisitor = std::function<void(const VariantKey&, Segment&&)>;
 
 class DuplicateKeyException : public ArcticSpecificException<ErrorCode::E_DUPLICATE_KEY> {
 public:
```

We get compilation errors like:

```cpp
  /home/jjerphan/dev/adb/cpp/arcticdb/storage/memory/memory_storage-inl.hpp:81:32: error: cannot bind rvalue reference of type 'arcticdb::Segment&&' to lvalue of type 'arcticdb::Segment'
     81 |                     visitor(k, seg);
        |                                ^~~
```

or like:

```cpp
  /home/jjerphan/dev/adb/cpp/arcticdb/storage/mongo/mongo_storage-inl.hpp:64:28: error: no match for call to '(const ReadVisitor {aka const std::function<void(const std::variant<arcticdb::entity::AtomKeyImpl, arcticdb::entity::RefKey>&, arcticdb::Segment&&)>}) (std::variant<arcticdb::entity::AtomKeyImpl, arcticdb::entity::RefKey>&, arcticdb::Segment&)'
     64 |                     visitor(k, kv.segment());
        |                     ~~~~~~~^~~~~~~~~~~~~~~~~
```

For instance, when we use l-value reference:

```diff
diff --git i/cpp/arcticdb/storage/storage.hpp w/cpp/arcticdb/storage/storage.hpp
index 48c71e7..a85f5b6 100644
--- i/cpp/arcticdb/storage/storage.hpp
+++ w/cpp/arcticdb/storage/storage.hpp
@@ -27,7 +27,7 @@ namespace arcticdb::storage {
 
 struct StorageBase {}; // marker class for type checking
 
-using ReadVisitor = std::function<void(const VariantKey&, Segment)>;
+using ReadVisitor = std::function<void(const VariantKey&, Segment&)>;
 
 class DuplicateKeyException : public ArcticSpecificException<ErrorCode::E_DUPLICATE_KEY> {
 public:
```

We get compilation errors like:
```cpp
  /home/jjerphan/dev/adb/cpp/arcticdb/storage/lmdb/lmdb_storage-inl.hpp:115:24: error: cannot bind non-const lvalue reference of type 'arcticdb::Segment&' to an rvalue of type 'arcticdb::Segment'
    115 |                 visitor(k, Segment::from_bytes(reinterpret_cast<std::uint8_t *>(mdb_val.mv_data),
        |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    116 |                                                mdb_val.mv_size));
        |                                                ~~~~~~~~~~~~~~~~~
```

or like:

```cpp
  /home/jjerphan/dev/adb/cpp/arcticdb/storage/s3/s3_storage-inl.hpp:262:24: error: no match for call to '(const ReadVisitor {aka const std::function<void(const std::variant<arcticdb::entity::AtomKeyImpl, arcticdb::entity::RefKey>&, arcticdb::Segment&)>}) (std::variant<arcticdb::entity::AtomKeyImpl, arcticdb::entity::RefKey>&, arcticdb::Segment)'
    262 |                 visitor(k, Segment::from_buffer(retrieved.get_buffer()));
        |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

</details>

Which adaptations should we perform?

Reply:
I think we should make a decision on what the ReadVisitor signature is, either a const reference or an r-value, but definitely not an r-value which will copy all the data and have a significant negative performance impact. Then we should change the call sites to conform to this pattern.

Personally I would prefer move, but I want to avoid at all costs taking this branch: https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/codec/segment.hpp#L240, which again copies the buffer. What I would suggest would be changing this line to throw, running the tests, and if the copy branch is not taken then we can use move everywhere for the read visitor.

Probably a lesson learned here in that it's better to be more restrictive about the signature required, so that we don't end up with a mess of slightly different ones.

Reply:
[`ReadVisitor` is now defined with rvalue references](https://github.com/man-group/ArcticDB/pull/670/files#diff-e9592f894f71949908c88db11096271fa492cdf2dda147ae4b2daf01cac640c1R30):

```cpp
using ReadVisitor = std::function<void(const VariantKey&, Segment &&)>;
```

Reply:
@jjerphan do the tests pass if the line William referenced is changed to just throw? If so, maybe we should make that change as well? Otherwise there will be a bear-trap where a seemingly cheap `std::move` is unexpectedly expensive.

Reply:
Thanks for the heads-up.

I am looking into it again now that rvalue references for `Segment` are used.

Reply:
I recompiled and reran the Python `unit` and `integration` test suites and C++ test suite (`test_unit_arcticdb`) with those changes:
```diff
diff --git i/cpp/arcticdb/codec/segment.hpp w/cpp/arcticdb/codec/segment.hpp
index e3e4b196..71922ae3 100644
--- i/cpp/arcticdb/codec/segment.hpp
+++ w/cpp/arcticdb/codec/segment.hpp
@@ -237,9 +237,7 @@ class Segment {
                 swap(std::get<BufferView>(buffer_), std::get<BufferView>(that.buffer_));
             }
         } else if (is_owning_buffer()) {
-            log::storage().info("Copying segment");
-            // data of segment being moved is not owned, moving it is dangerous, copying instead
-            that.buffer().copy_to(*std::get<std::shared_ptr<Buffer>>(buffer_));
+            throw std::runtime_error("Moving segment with owning buffer to segment with non-owning buffer");
         } else {
             // data of this segment is a view, but the move data is moved
             buffer_ = std::move(std::get<std::shared_ptr<Buffer>>(that.buffer_));
```

and I did not observe any difference: all C++ tests pass and all Python test pass except the ones currently failing on `master`.